File: js/src/tools.js
import { SCHEMA_VERSION, SCHEMA_VERSION_V2 } from "./constants.js";
import { convertUECv1toV2 } from "./convert.js";
import { validateUEC, isUEC } from "./validate.js";
import { isPlainObject } from "./utils.js";

const isAssetLocatorObject = (value) =>
  isPlainObject(value) &&
  (value.type === "inline_base64" ||
    value.type === "remote_url" ||
    value.type === "asset_ref");

const isLikelyAssetString = (value) =>
  typeof value === "string" &&
  (value.startsWith("http://") ||
    value.startsWith("https://") ||
    value.startsWith("data:"));

const normalizeValue = (value) => {
  if (Array.isArray(value)) {
    return value.map((item) => normalizeValue(item));
  }

  if (!isPlainObject(value)) {
    return value;
  }

  const out = {};
  Object.keys(value)
    .sort()
    .forEach((key) => {
      if (value[key] === undefined) {
        return;
      }
      out[key] = normalizeValue(value[key]);
    });

  return out;
};

const deepEqual = (a, b) => {
  if (a === b) {
    return true;
  }

  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) {
      return false;
    }
    for (let i = 0; i < a.length; i += 1) {
      if (!deepEqual(a[i], b[i])) {
        return false;
      }
    }
    return true;
  }

  if (isPlainObject(a) && isPlainObject(b)) {
    const keysA = Object.keys(a);
    const keysB = Object.keys(b);
    if (keysA.length !== keysB.length) {
      return false;
    }
    for (const key of keysA) {
      if (!Object.hasOwn(b, key) || !deepEqual(a[key], b[key])) {
        return false;
      }
    }
    return true;
  }

  return false;
};

const walkDiff = (a, b, path, out) => {
  if (deepEqual(a, b)) {
    return;
  }

  if (Array.isArray(a) && Array.isArray(b)) {
    const max = Math.max(a.length, b.length);
    for (let i = 0; i < max; i += 1) {
      walkDiff(a[i], b[i], `${path}[${i}]`, out);
    }
    return;
  }

  if (isPlainObject(a) && isPlainObject(b)) {
    const keys = new Set([...Object.keys(a), ...Object.keys(b)]);
    Array.from(keys)
      .sort()
      .forEach((key) => {
        const nextPath = path ? `${path}.${key}` : key;
        if (!Object.hasOwn(a, key)) {
          out.push({ path: nextPath, type: "added", after: b[key] });
          return;
        }
        if (!Object.hasOwn(b, key)) {
          out.push({ path: nextPath, type: "removed", before: a[key] });
          return;
        }
        walkDiff(a[key], b[key], nextPath, out);
      });
    return;
  }

  out.push({ path: path || "root", type: "changed", before: a, after: b });
};

const deepClone = (value) => {
  if (Array.isArray(value)) {
    return value.map((item) => deepClone(item));
  }
  if (!isPlainObject(value)) {
    return value;
  }
  const out = {};
  for (const key of Object.keys(value)) {
    out[key] = deepClone(value[key]);
  }
  return out;
};

const mergeValues = (base, incoming, path, strategy, conflicts) => {
  if (incoming === undefined) {
    return deepClone(base);
  }

  if (base === undefined) {
    return deepClone(incoming);
  }

  if (Array.isArray(base) && Array.isArray(incoming)) {
    if (strategy.array === "concat") {
      return [...deepClone(base), ...deepClone(incoming)];
    }
    if (!deepEqual(base, incoming)) {
      conflicts.push(path || "root");
    }
    return deepClone(incoming);
  }

  if (isPlainObject(base) && isPlainObject(incoming)) {
    const out = {};
    const keys = new Set([...Object.keys(base), ...Object.keys(incoming)]);
    for (const key of keys) {
      const nextPath = path ? `${path}.${key}` : key;
      out[key] = mergeValues(base[key], incoming[key], nextPath, strategy, conflicts);
    }
    return out;
  }

  if (!deepEqual(base, incoming)) {
    conflicts.push(path || "root");
  }

  return strategy.conflict === "base" ? deepClone(base) : deepClone(incoming);
};

const toV1FromV2 = (card, options = {}) => {
  const warnings = [];
  const next = deepClone(card);

  next.schema = { ...next.schema, version: SCHEMA_VERSION };

  if (isPlainObject(next.payload)) {
    const payload = { ...next.payload };

    if (isPlainObject(payload.scene)) {
      const scene = { ...payload.scene };
      if (Object.hasOwn(scene, "selectedVariant")) {
        if (scene.selectedVariant === 0) {
          scene.selectedVariantId = null;
        } else if (typeof scene.selectedVariant === "string") {
          scene.selectedVariantId = scene.selectedVariant;
        }
        delete scene.selectedVariant;
      }
      payload.scenes = [scene];
      payload.defaultSceneId = scene.id;
      delete payload.scene;
    }

    if (payload.promptTemplateId !== undefined) {
      if (!payload.systemPrompt || payload.systemPrompt === null) {
        payload.systemPrompt = `_ID:${payload.promptTemplateId}`;
      }
      warnings.push("payload.promptTemplateId was mapped to v1 systemPrompt and then removed");
      delete payload.promptTemplateId;
    }

    const droppedV2Fields = [
      "fallbackModelId",
      "nickname",
      "creator",
      "creatorNotes",
      "creatorNotesMultilingual",
      "source",
      "characterBook",
    ];

    for (const field of droppedV2Fields) {
      if (payload[field] !== undefined) {
        warnings.push(`payload.${field} is not supported in v1 and was removed`);
        delete payload[field];
      }
    }

    if (!options.keepRules && payload.rules === undefined) {
      payload.rules = [];
    }

    next.payload = payload;
  }

  if (isPlainObject(next.meta)) {
    const meta = { ...next.meta };
    if (meta.originalCreatedAt !== undefined) {
      delete meta.originalCreatedAt;
      warnings.push("meta.originalCreatedAt was removed for v1 compatibility");
    }
    if (meta.originalUpdatedAt !== undefined) {
      delete meta.originalUpdatedAt;
      warnings.push("meta.originalUpdatedAt was removed for v1 compatibility");
    }
    if (meta.originalSource !== undefined) {
      delete meta.originalSource;
      warnings.push("meta.originalSource was removed for v1 compatibility");
    }
    next.meta = meta;
  }

  return { card: next, warnings };
};

export const parseUEC = (json, options) => {
  if (typeof json !== "string") {
    return { ok: false, errors: ["root: json input must be a string"] };
  }

  let parsed;
  try {
    parsed = JSON.parse(json);
  } catch (error) {
    return { ok: false, errors: [`root: invalid JSON (${error.message})`] };
  }

  const result = validateUEC(parsed, options);
  if (!result.ok) {
    return result;
  }

  return { ok: true, value: parsed, errors: [] };
};

export const stringifyUEC = (card, options = {}) => {
  const { space = 2 } = options;
  const normalized = normalizeUEC(card);
  return JSON.stringify(normalized, null, space);
};

export const normalizeUEC = (card) => {
  const out = normalizeValue(card);

  if (isPlainObject(out) && !isPlainObject(out.app_specific_settings)) {
    out.app_specific_settings = {};
  }
  if (isPlainObject(out) && !isPlainObject(out.meta)) {
    out.meta = {};
  }
  if (isPlainObject(out) && !isPlainObject(out.extensions)) {
    out.extensions = {};
  }

  return out;
};

export const upgradeUEC = (card, targetVersion = SCHEMA_VERSION_V2) => {
  if (!isPlainObject(card) || !isPlainObject(card.schema)) {
    throw new Error("card must be an object with a schema");
  }

  if (targetVersion === SCHEMA_VERSION_V2) {
    if (card.schema.version === SCHEMA_VERSION_V2) {
      return normalizeUEC(card);
    }
    if (card.schema.version === SCHEMA_VERSION) {
      return convertUECv1toV2(card);
    }
    throw new Error(`unsupported source version: ${String(card.schema.version)}`);
  }

  if (targetVersion === SCHEMA_VERSION) {
    return downgradeUEC(card, targetVersion).card;
  }

  throw new Error(`unsupported target version: ${String(targetVersion)}`);
};

export const downgradeUEC = (card, targetVersion = SCHEMA_VERSION, options = {}) => {
  if (!isPlainObject(card) || !isPlainObject(card.schema)) {
    throw new Error("card must be an object with a schema");
  }

  if (targetVersion !== SCHEMA_VERSION) {
    throw new Error(`unsupported target version: ${String(targetVersion)}`);
  }

  if (card.schema.version === SCHEMA_VERSION) {
    return { card: normalizeUEC(card), warnings: [] };
  }

  if (card.schema.version !== SCHEMA_VERSION_V2) {
    throw new Error(`unsupported source version: ${String(card.schema.version)}`);
  }

  return toV1FromV2(card, options);
};

export const diffUEC = (a, b) => {
  const left = normalizeUEC(a);
  const right = normalizeUEC(b);
  const changes = [];
  walkDiff(left, right, "", changes);
  return changes;
};

export const mergeUEC = (base, incoming, strategy = {}) => {
  const settings = {
    array: strategy.array === "concat" ? "concat" : "replace",
    conflict: strategy.conflict === "base" ? "base" : "incoming",
  };

  const conflicts = [];
  const value = mergeValues(base, incoming, "", settings, conflicts);
  return { value, conflicts: Array.from(new Set(conflicts)).sort() };
};

export const validateUECStrict = (value, options = {}) =>
  validateUEC(value, { ...options, strict: true });

export const validateUECAtVersion = (value, version, options = {}) => {
  const result = validateUEC(value, options);

  if (!isPlainObject(value) || !isPlainObject(value.schema)) {
    return result;
  }

  if (value.schema.version !== version) {
    return {
      ok: false,
      errors: [
        ...result.errors,
        `schema.version: expected \"${version}\" but received \"${value.schema.version}\"`,
      ],
    };
  }

  return result;
};

export const isCharacterUEC = (value, options) =>
  isUEC(value, options) && value.kind === "character";

export const isPersonaUEC = (value, options) =>
  isUEC(value, options) && value.kind === "persona";

export const extractAssets = (card) => {
  const assets = [];

  const walk = (value, path) => {
    if (isLikelyAssetString(value)) {
      assets.push({ path, kind: "string", value });
      return;
    }

    if (isAssetLocatorObject(value)) {
      assets.push({ path, kind: "locator", value: deepClone(value) });
      return;
    }

    if (Array.isArray(value)) {
      value.forEach((item, index) => walk(item, `${path}[${index}]`));
      return;
    }

    if (!isPlainObject(value)) {
      return;
    }

    Object.keys(value).forEach((key) => {
      const nextPath = path ? `${path}.${key}` : key;
      walk(value[key], nextPath);
    });
  };

  walk(card, "");
  return assets;
};

export const rewriteAssets = (card, mapper) => {
  const rewrite = (value, path) => {
    if (isLikelyAssetString(value)) {
      return mapper({ path, kind: "string", value });
    }

    if (isAssetLocatorObject(value)) {
      return mapper({ path, kind: "locator", value: deepClone(value) });
    }

    if (Array.isArray(value)) {
      return value.map((item, index) => rewrite(item, `${path}[${index}]`));
    }

    if (!isPlainObject(value)) {
      return value;
    }

    const out = {};
    Object.keys(value).forEach((key) => {
      const nextPath = path ? `${path}.${key}` : key;
      out[key] = rewrite(value[key], nextPath);
    });
    return out;
  };

  return rewrite(card, "");
};

export const lintUEC = (card) => {
  const warnings = [];

  if (!isPlainObject(card) || !isPlainObject(card.payload)) {
    return { ok: false, warnings: ["root: not a valid UEC object shape"] };
  }

  if (typeof card.payload.description === "string" && card.payload.description.trim() === "") {
    warnings.push("payload.description is an empty string");
  }

  if (
    typeof card.payload.createdAt === "number" &&
    typeof card.payload.updatedAt === "number" &&
    card.payload.createdAt > card.payload.updatedAt
  ) {
    warnings.push("payload.createdAt is greater than payload.updatedAt");
  }

  if (isPlainObject(card.meta)) {
    if (
      typeof card.meta.createdAt === "number" &&
      typeof card.meta.updatedAt === "number" &&
      card.meta.createdAt > card.meta.updatedAt
    ) {
      warnings.push("meta.createdAt is greater than meta.updatedAt");
    }
  }

  if (
    isPlainObject(card.schema) &&
    card.schema.version === SCHEMA_VERSION_V2 &&
    isPlainObject(card.payload.scene) &&
    typeof card.payload.scene.selectedVariant === "string" &&
    Array.isArray(card.payload.scene.variants)
  ) {
    const ids = new Set(
      card.payload.scene.variants
        .filter((variant) => isPlainObject(variant) && typeof variant.id === "string")
        .map((variant) => variant.id),
    );

    if (!ids.has(card.payload.scene.selectedVariant)) {
      warnings.push("payload.scene.selectedVariant does not match any variant id");
    }
  }

  extractAssets(card).forEach((asset) => {
    if (
      asset.kind === "locator" &&
      asset.value.type === "inline_base64" &&
      typeof asset.value.data === "string" &&
      asset.value.data.length > 200000
    ) {
      warnings.push(`${asset.path}: inline_base64 asset is very large`);
    }
  });

  return { ok: warnings.length === 0, warnings };
};

File: rust/src/tools.rs
use serde_json::{Map, Value};
use std::collections::BTreeSet;

use crate::constants::{SCHEMA_VERSION, SCHEMA_VERSION_V2};
use crate::convert::convert_uec_v1_to_v2;
use crate::types::{
    AssetReference, DowngradeResult, LintResult, MergeOptions, MergeResult, ParseValidationResult,
    UecDiffEntry,
};
use crate::utils::{is_asset_locator_object, is_likely_asset_string, is_object, normalize_value};
use crate::validators::validate_uec;

pub fn normalize_uec(card: &Value) -> Value {
    let mut normalized = normalize_value(card);

    if let Some(root) = normalized.as_object_mut() {
        if !root.get("app_specific_settings").is_some_and(is_object) {
            root.insert(
                "app_specific_settings".to_string(),
                Value::Object(Map::new()),
            );
        }

        if !root.get("meta").is_some_and(is_object) {
            root.insert("meta".to_string(), Value::Object(Map::new()));
        }

        if !root.get("extensions").is_some_and(is_object) {
            root.insert("extensions".to_string(), Value::Object(Map::new()));
        }
    }

    normalized
}

pub fn parse_uec(input: &str, strict: bool) -> ParseValidationResult {
    let parsed = match serde_json::from_str::<Value>(input) {
        Ok(value) => value,
        Err(error) => {
            return ParseValidationResult {
                ok: false,
                value: None,
                errors: vec![format!("root: invalid JSON ({})", error)],
            };
        }
    };

    let result = validate_uec(&parsed, strict);
    if result.ok {
        ParseValidationResult {
            ok: true,
            value: Some(parsed),
            errors: Vec::new(),
        }
    } else {
        ParseValidationResult {
            ok: false,
            value: None,
            errors: result.errors,
        }
    }
}

pub fn stringify_uec(card: &Value, pretty: bool) -> Result<String, String> {
    let normalized = normalize_uec(card);
    if pretty {
        serde_json::to_string_pretty(&normalized).map_err(|err| err.to_string())
    } else {
        serde_json::to_string(&normalized).map_err(|err| err.to_string())
    }
}

pub fn upgrade_uec(card: &Value, target_version: &str) -> Result<Value, String> {
    let version = card
        .get("schema")
        .and_then(|schema| schema.get("version"))
        .and_then(Value::as_str)
        .ok_or_else(|| "card must be an object with a schema".to_string())?;

    match target_version {
        SCHEMA_VERSION_V2 => {
            if version == SCHEMA_VERSION_V2 {
                Ok(normalize_uec(card))
            } else if version == SCHEMA_VERSION {
                convert_uec_v1_to_v2(card)
            } else {
                Err(format!("unsupported source version: {}", version))
            }
        }
        SCHEMA_VERSION => downgrade_uec(card, SCHEMA_VERSION, false).map(|result| result.card),
        _ => Err(format!("unsupported target version: {}", target_version)),
    }
}

pub fn downgrade_uec(
    card: &Value,
    target_version: &str,
    keep_rules: bool,
) -> Result<DowngradeResult, String> {
    if target_version != SCHEMA_VERSION {
        return Err(format!("unsupported target version: {}", target_version));
    }

    let version = card
        .get("schema")
        .and_then(|schema| schema.get("version"))
        .and_then(Value::as_str)
        .ok_or_else(|| "card must be an object with a schema".to_string())?;

    if version == SCHEMA_VERSION {
        return Ok(DowngradeResult {
            card: normalize_uec(card),
            warnings: Vec::new(),
        });
    }

    if version != SCHEMA_VERSION_V2 {
        return Err(format!("unsupported source version: {}", version));
    }

    let mut warnings = Vec::new();
    let mut next = card.clone();

    if let Some(schema) = next.get_mut("schema").and_then(Value::as_object_mut) {
        schema.insert(
            "version".to_string(),
            Value::String(SCHEMA_VERSION.to_string()),
        );
    }

    if let Some(payload) = next.get_mut("payload").and_then(Value::as_object_mut) {
        if let Some(scene) = payload.remove("scene")
            && let Value::Object(mut scene_map) = scene
        {
            if let Some(selected) = scene_map.remove("selectedVariant") {
                if matches!(selected, Value::Number(ref number) if number.as_i64() == Some(0)) {
                    scene_map.insert("selectedVariantId".to_string(), Value::Null);
                } else {
                    scene_map.insert("selectedVariantId".to_string(), selected);
                }
            }

            let scene_id = scene_map.get("id").cloned();
            payload.insert(
                "scenes".to_string(),
                Value::Array(vec![Value::Object(scene_map)]),
            );
            if let Some(id) = scene_id {
                payload.insert("defaultSceneId".to_string(), id);
            }
        }

        if let Some(prompt_template_id) = payload.remove("promptTemplateId") {
            if payload.get("systemPrompt").is_none_or(Value::is_null)
                && let Some(template_id) = prompt_template_id.as_str()
            {
                payload.insert(
                    "systemPrompt".to_string(),
                    Value::String(format!("_ID:{}", template_id)),
                );
            }
            warnings.push(
                "payload.promptTemplateId was mapped to v1 systemPrompt and then removed"
                    .to_string(),
            );
        }

        let removed_fields = [
            "fallbackModelId",
            "nickname",
            "creator",
            "creatorNotes",
            "creatorNotesMultilingual",
            "source",
            "characterBook",
        ];

        for field in removed_fields {
            if payload.remove(field).is_some() {
                warnings.push(format!(
                    "payload.{} is not supported in v1 and was removed",
                    field
                ));
            }
        }

        if !keep_rules && !payload.contains_key("rules") {
            payload.insert("rules".to_string(), Value::Array(Vec::new()));
        }
    }

    if let Some(meta) = next.get_mut("meta").and_then(Value::as_object_mut) {
        if meta.remove("originalCreatedAt").is_some() {
            warnings.push("meta.originalCreatedAt was removed for v1 compatibility".to_string());
        }
        if meta.remove("originalUpdatedAt").is_some() {
            warnings.push("meta.originalUpdatedAt was removed for v1 compatibility".to_string());
        }
        if meta.remove("originalSource").is_some() {
            warnings.push("meta.originalSource was removed for v1 compatibility".to_string());
        }
    }

    Ok(DowngradeResult {
        card: next,
        warnings,
    })
}

fn walk_diff(a: &Value, b: &Value, path: &str, out: &mut Vec<UecDiffEntry>) {
    if a == b {
        return;
    }

    match (a, b) {
        (Value::Array(left), Value::Array(right)) => {
            let max_len = left.len().max(right.len());
            for index in 0..max_len {
                let next_path = format!("{}[{}]", path, index);
                let left_value = left.get(index).unwrap_or(&Value::Null);
                let right_value = right.get(index).unwrap_or(&Value::Null);
                walk_diff(left_value, right_value, &next_path, out);
            }
        }
        (Value::Object(left), Value::Object(right)) => {
            let mut keys = BTreeSet::new();
            keys.extend(left.keys().cloned());
            keys.extend(right.keys().cloned());

            for key in keys {
                let next_path = if path.is_empty() {
                    key.clone()
                } else {
                    format!("{}.{}", path, key)
                };

                match (left.get(&key), right.get(&key)) {
                    (None, Some(after)) => out.push(UecDiffEntry {
                        path: next_path,
                        change_type: "added".to_string(),
                        before: None,
                        after: Some(after.clone()),
                    }),
                    (Some(before), None) => out.push(UecDiffEntry {
                        path: next_path,
                        change_type: "removed".to_string(),
                        before: Some(before.clone()),
                        after: None,
                    }),
                    (Some(before), Some(after)) => walk_diff(before, after, &next_path, out),
                    _ => {}
                }
            }
        }
        _ => out.push(UecDiffEntry {
            path: if path.is_empty() {
                "root".to_string()
            } else {
                path.to_string()
            },
            change_type: "changed".to_string(),
            before: Some(a.clone()),
            after: Some(b.clone()),
        }),
    }
}

pub fn diff_uec(left: &Value, right: &Value) -> Vec<UecDiffEntry> {
    let a = normalize_uec(left);
    let b = normalize_uec(right);
    let mut changes = Vec::new();
    walk_diff(&a, &b, "", &mut changes);
    changes
}

fn merge_values(
    base: &Value,
    incoming: &Value,
    path: &str,
    options: &MergeOptions,
    conflicts: &mut BTreeSet<String>,
) -> Value {
    if incoming.is_null() {
        return base.clone();
    }

    match (base, incoming) {
        (Value::Array(left), Value::Array(right)) => {
            if options.array.as_deref() == Some("concat") {
                let mut merged = left.clone();
                merged.extend(right.clone());
                Value::Array(merged)
            } else {
                if left != right {
                    conflicts.insert(path.to_string());
                }
                Value::Array(right.clone())
            }
        }
        (Value::Object(left), Value::Object(right)) => {
            let mut merged = Map::new();
            let mut keys = BTreeSet::new();
            keys.extend(left.keys().cloned());
            keys.extend(right.keys().cloned());

            for key in keys {
                let next_path = if path.is_empty() {
                    key.clone()
                } else {
                    format!("{}.{}", path, key)
                };

                let left_value = left.get(&key).unwrap_or(&Value::Null);
                let right_value = right.get(&key).unwrap_or(&Value::Null);

                let merged_value = if right.contains_key(&key) {
                    merge_values(left_value, right_value, &next_path, options, conflicts)
                } else {
                    left_value.clone()
                };

                merged.insert(key, merged_value);
            }

            Value::Object(merged)
        }
        _ => {
            if base != incoming {
                conflicts.insert(path.to_string());
            }

            if options.conflict.as_deref() == Some("base") {
                base.clone()
            } else {
                incoming.clone()
            }
        }
    }
}

pub fn merge_uec(base: &Value, incoming: &Value, options: MergeOptions) -> MergeResult {
    let mut conflicts = BTreeSet::new();
    let merged = merge_values(base, incoming, "", &options, &mut conflicts);
    MergeResult {
        value: merged,
        conflicts: conflicts
            .into_iter()
            .filter(|item| !item.is_empty())
            .collect(),
    }
}

fn extract_assets_walk(value: &Value, path: &str, assets: &mut Vec<AssetReference>) {
    if is_likely_asset_string(value) {
        assets.push(AssetReference {
            path: path.to_string(),
            kind: "string".to_string(),
            value: value.clone(),
        });
        return;
    }

    if is_object(value) && is_asset_locator_object(value) {
        assets.push(AssetReference {
            path: path.to_string(),
            kind: "locator".to_string(),
            value: value.clone(),
        });
        return;
    }

    match value {
        Value::Array(items) => {
            for (index, item) in items.iter().enumerate() {
                let next_path = format!("{}[{}]", path, index);
                extract_assets_walk(item, &next_path, assets);
            }
        }
        Value::Object(map) => {
            for (key, item) in map {
                let next_path = if path.is_empty() {
                    key.clone()
                } else {
                    format!("{}.{}", path, key)
                };
                extract_assets_walk(item, &next_path, assets);
            }
        }
        _ => {}
    }
}

pub fn extract_assets(card: &Value) -> Vec<AssetReference> {
    let mut assets = Vec::new();
    extract_assets_walk(card, "", &mut assets);
    assets
}

pub fn rewrite_assets<F>(card: &Value, mapper: &mut F) -> Value
where
    F: FnMut(AssetReference) -> Value,
{
    fn walk<F>(value: &Value, path: &str, mapper: &mut F) -> Value
    where
        F: FnMut(AssetReference) -> Value,
    {
        if is_likely_asset_string(value) {
            return mapper(AssetReference {
                path: path.to_string(),
                kind: "string".to_string(),
                value: value.clone(),
            });
        }

        if is_object(value) && is_asset_locator_object(value) {
            return mapper(AssetReference {
                path: path.to_string(),
                kind: "locator".to_string(),
                value: value.clone(),
            });
        }

        match value {
            Value::Array(items) => Value::Array(
                items
                    .iter()
                    .enumerate()
                    .map(|(index, item)| walk(item, &format!("{}[{}]", path, index), mapper))
                    .collect(),
            ),
            Value::Object(map) => {
                let mut out = Map::new();
                for (key, item) in map {
                    let next_path = if path.is_empty() {
                        key.clone()
                    } else {
                        format!("{}.{}", path, key)
                    };
                    out.insert(key.clone(), walk(item, &next_path, mapper));
                }
                Value::Object(out)
            }
            _ => value.clone(),
        }
    }

    walk(card, "", mapper)
}

pub fn lint_uec(card: &Value) -> LintResult {
    let mut warnings = Vec::new();

    let Some(payload) = card.get("payload").and_then(Value::as_object) else {
        return LintResult {
            ok: false,
            warnings: vec!["root: not a valid UEC object shape".to_string()],
        };
    };

    if payload
        .get("description")
        .and_then(Value::as_str)
        .is_some_and(|description| description.trim().is_empty())
    {
        warnings.push("payload.description is an empty string".to_string());
    }

    if payload
        .get("createdAt")
        .and_then(Value::as_i64)
        .zip(payload.get("updatedAt").and_then(Value::as_i64))
        .is_some_and(|(created, updated)| created > updated)
    {
        warnings.push("payload.createdAt is greater than payload.updatedAt".to_string());
    }

    if card
        .get("meta")
        .and_then(Value::as_object)
        .and_then(|meta| {
            meta.get("createdAt")
                .and_then(Value::as_i64)
                .zip(meta.get("updatedAt").and_then(Value::as_i64))
        })
        .is_some_and(|(created, updated)| created > updated)
    {
        warnings.push("meta.createdAt is greater than meta.updatedAt".to_string());
    }

    if card
        .get("schema")
        .and_then(|schema| schema.get("version"))
        .and_then(Value::as_str)
        == Some(SCHEMA_VERSION_V2)
        && let Some(scene) = payload.get("scene").and_then(Value::as_object)
        && let Some(selected_variant) = scene.get("selectedVariant").and_then(Value::as_str)
        && let Some(variants) = scene.get("variants").and_then(Value::as_array)
    {
        let known_ids: BTreeSet<String> = variants
            .iter()
            .filter_map(|variant| {
                variant
                    .as_object()
                    .and_then(|map| map.get("id"))
                    .and_then(Value::as_str)
                    .map(ToOwned::to_owned)
            })
            .collect();

        if !known_ids.contains(selected_variant) {
            warnings
                .push("payload.scene.selectedVariant does not match any variant id".to_string());
        }
    }

    for asset in extract_assets(card) {
        if asset.kind == "locator"
            && asset.value.get("type").and_then(Value::as_str) == Some("inline_base64")
            && asset
                .value
                .get("data")
                .and_then(Value::as_str)
                .is_some_and(|data| data.len() > 200_000)
        {
            warnings.push(format!("{}: inline_base64 asset is very large", asset.path));
        }
    }

    LintResult {
        ok: warnings.is_empty(),
        warnings,
    }
}
